[nnc] Fix half2float conversion and re-enable float16

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64196
* __->__ #64147

Fused float16 operations are cast to float for arithmetic, and then cast back down to float16 for storage, since (1) fused kernels tend to be bandwith bound anyways, and (2) some older gpus had pretty bad fp16 perf, and (3) this change actually improves numeric accuracy.

The problem for NNC was that we weren't properly converting all the intermediate ops; we were just leaving them all as "half" and hoping that (because the CUDA C++ we spit out doesn't usually name the types explicitly) that it would work out OK.  It sort of does, but weird bugs crop up, like the final cast back to half (`result[...] = half(...expr...);` would get "optimized" away, because the IR already thought the result was half.

This change properly converts all the float16 arithmetic in a kernel to float32 in the IR, getting rid of this class of weird bugs.

bertmaher committed:
[nnc] Fix half2float conversion and re-enable float16

[ghstack-poisoned]

bertmaher commented:
We disabled half in the fuser because of issues like #61382.  This PR fixes some remaining issues in half-rewriting and re-enables fp16 in the fuser.

bertmaher committed:
Update on "[nnc] Fix half2float conversion and re-enable float16"




[ghstack-poisoned]

ZolotukhinM commented on test/test_jit_fuser_te.py:
```diff
@@ -95,7 +96,7 @@ def setUp(self):
         ]
         self.fp_dtypes = [
             # TODO: Add back when https://github.com/pytorch/pytorch/issues/55905 is closed
```
Nit: remove TODO

ZolotukhinM commented on test/test_jit_fuser_te.py:
```diff
@@ -1768,7 +1790,9 @@ def test_type_as_cat(self):
                 return torch.cat((x, y.type_as(x)), dim=1)
-            for dtype1, dtype2 in product(self.dtypes, self.dtypes):
+            dtypes = self.dtypes.copy()
+            dtypes.remove(torch.float16)
```
Nit: might be worth of a comment?

ZolotukhinM commented on test/test_jit_fuser_te.py:
```diff
@@ -1936,6 +1960,22 @@ def bn_neither(i, x):
             for fn in [bn, bn_no_weight, bn_no_bias, bn_neither]:
                 test(fn, (i, x))
 
+    @unittest.skipIf(not RUN_CUDA, "fuser requires CUDA for fp16")
+    def test_bn_half_61382(self):
```
We already have an equivalent test, it's probably better to just re-enable it:
https://github.com/pytorch/pytorch/blob/44e3ed88c9a1bd9ee6b0168ba5271a2c6b006cc8/test/test_tensorexpr.py#L1225-L1241

ZolotukhinM commented on torch/csrc/jit/tensorexpr/ir_verifier.cpp:
```diff
@@ -119,7 +119,19 @@ void IRVerifier::visit(IfThenElsePtr v) {
+    return;
+  }
   // TODO: add a check for OpArgCount and op_type
+  for (auto const& param : v->params()) {
+    if (param->dtype() != v->dtype()) {
+      throw malformed_ir("bad dtype in intrinsic arg");
+    }
+  }
```
Is it a requirement that all parameters should have the same dtype as the result value? I guess it's true at the moment, but I wonder if it's appropriate to enforce this condition in IR verifier

bertmaher replied:
Yeah it's definitely not universally true (as you see I had to special-case `isnan`), but it's otherwise true for the set of intrinsics we have.  I think llvm would complain anywhere this doesn't hold, since the type signatures wouldn't match.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/half_support.h:
```diff
@@ -128,6 +128,46 @@ class HalfRewriter : public IRMutator {
+    return v;
+  }
+
+  ExprPtr mutate(AddPtr v) override {
```
Do you mind adding a brief description of what was broken in that test and how this change fixes it?

bertmaher replied:
Sure thing, will update the summary.

ZolotukhinM reviewed (approved):
Looks good modulo minor comments!

bertmaher committed:
Update on "[nnc] Fix half2float conversion and re-enable float16"


Fused float16 operations are cast to float for arithmetic, and then cast back down to float16 for storage, since (1) fused kernels tend to be bandwith bound anyways, and (2) some older gpus had pretty bad fp16 perf, and (3) this change actually improves numeric accuracy.

The problem for NNC was that we weren't properly converting all the intermediate ops; we were just leaving them all as "half" and hoping that (because the CUDA C++ we spit out doesn't usually name the types explicitly) that it would work out OK.  It sort of does, but weird bugs crop up, like the final cast back to half (`result[...] = half(...expr...);` would get "optimized" away, because the IR already thought the result was half.

This change properly converts all the float16 arithmetic in a kernel to float32 in the IR, getting rid of this class of weird bugs.

[ghstack-poisoned]
